lighthouse-core/audits/lazy-third-party.js

```diff
@@ -0,0 +1,198 @@
+  /** Summary text for the result of a Lighthouse audit that identifies the third party code on a web page that can be lazy loaded. This text summarizes the number of lazy loading facades that can be used on the page. Lazy loading means loading resources is deferred until they are needed. */
+  displayValue: `{itemCount, plural,
+  =1 {# facade alternative available}
+  other {# facade alternatives available}
+  }`,
+  /** Label for a table column that displays the name of a lazy loading facade alternative for a third party resource. Lazy loading means loading resources is deferred until they are needed. */
+  columnFacade: 'Facade Alternative',
+  /** Label for a table column that displays the name of the third party product that a URL is used for. */
+  columnProduct: 'Product',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/** @typedef {import("third-party-web").IEntity} ThirdPartyEntity */
+/** @typedef {import("third-party-web").IProduct} ThirdPartyProduct*/
+
+/** @typedef {{product: ThirdPartyProduct, cutoffTime: number, urlSummaries: Map<string, ThirdPartySummary.Summary>}} ProductSummary */
+
+class LazyThirdParty extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'lazy-third-party',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['traces', 'devtoolsLogs', 'URL'],
+    };
+  }
+
+  /**
+   * @param {Map<string, ThirdPartySummary.Summary>} byURL
+   * @param {ThirdPartyEntity | undefined} mainEntity
+   * @return {ProductSummary[]}
+   */
+  static getProductSummaries(byURL, mainEntity) {
+    /** @type {Map<string, Map<string, ProductSummary>>} */
+    const entitySummaries = new Map();
+
+    for (const [url, urlSummary] of byURL) {
+      const entity = thirdPartyWeb.getEntity(url);
+      const product = thirdPartyWeb.getProduct(url);
+      if (!entity || thirdPartyWeb.isFirstParty(url, mainEntity)) continue;
+
+      /** @type {Map<string, ProductSummary>} */
+      const productSummaries = entitySummaries.get(entity.name) || new Map();
+      if (product && product.facades && product.facades.length) {
+        // Record new url if product has a facade.
+        const productSummary = productSummaries.get(product.name) || {
+          product,
+          cutoffTime: Infinity,
+          urlSummaries: new Map(),
+        };
+
+        productSummary.urlSummaries.set(url, urlSummary);
+        productSummary.cutoffTime = Math.min(productSummary.cutoffTime, urlSummary.endTime);
+
+        productSummaries.set(product.name, productSummary);
+      } else {
+        // If the url does not have a facade but one or more products on its entity do,
+        // we still want to record this url because it was probably fetched by a product with a facade.
+        for (const productSummary of productSummaries.values()) {
```

Comment:
is this assuming an ordering of `byURL` that if this url was part of a product that a summary for it would already exist in `productSummaries`?

Reply:
Good point, this should definitely be changed

Reply:
Interestingly, I noticed the results were the same after changing this (i.e. we weren't missing any URLs that happened to appear before the product URL in `byURL`.

I didn't know that `Map()` remembers to order of insertion. The product URL will always appear first in the network records so the entries will always be in an order where the `productSummary` is created first.

Reply:
> I didn't know that `Map()` remembers to order of insertion. The product URL will always appear first in the network records so the entries will always be in an order where the `productSummary` is created first.

Ah, got it, but that does assume network records are sorted correctly. That's probably usually true, but they are created from DevTools events and I don't think we ever explicitly sort them, so you never know.

This could sort `byURL` before this loop or this could take two passes (generate productSummaries, then associate URL with them), but that might be complete overkill. cc @patrickhulce to provide a check on my possibly worrying about nothing :)

Reply:
I have a solution that does two passes from when I was testing this, I don't think it will hurt to use it either way.